internal/eventstream/api_responses.go

```diff
@@ -0,0 +1,261 @@
+	"time"
+)
+
+// Icinga2Time is a custom time.Time type for JSON unmarshalling from Icinga 2's unix timestamp type.
+type Icinga2Time struct {
+	time.Time
+}
+
+func (iciTime *Icinga2Time) UnmarshalJSON(data []byte) error {
+	unixTs, err := strconv.ParseFloat(string(data), 64)
+	if err != nil {
+		return err
+	}
+
+	unixMicro := int64(unixTs * 1_000_000)
+	iciTime.Time = time.UnixMicro(unixMicro)
+	return nil
+}
+
+// Comment represents the Icinga 2 API Comment object.
+//
+// NOTE:
+//   - An empty Service field indicates a host comment.
+//   - The optional EntryType should be User = 1, Downtime = 2, Flapping = 3, Acknowledgement = 4.
+//
+// https://icinga.com/docs/icinga-2/latest/doc/09-object-types/#objecttype-comment
+type Comment struct {
+	Host      string      `json:"host_name"`
+	Service   string      `json:"service_name"`
+	Author    string      `json:"author"`
+	Text      string      `json:"text"`
+	EntryTime Icinga2Time `json:"entry_time"`
+	EntryType int         `json:"entry_type"`
+}
+
+// CheckResult represents the Icinga 2 API CheckResult object.
+//
+// https://icinga.com/docs/icinga-2/latest/doc/08-advanced-topics/#advanced-value-types-checkresult
+type CheckResult struct {
+	ExitStatus     int         `json:"exit_status"`
+	Output         string      `json:"output"`
+	State          int         `json:"state"`
+	ExecutionStart Icinga2Time `json:"execution_start"`
+	ExecutionEnd   Icinga2Time `json:"execution_end"`
+}
+
+// Downtime represents the Icinga 2 API Downtime object.
+//
+// NOTE:
+//   - An empty Service field indicates a host service.
+//
+// https://icinga.com/docs/icinga-2/latest/doc/09-object-types/#objecttype-downtime
+type Downtime struct {
+	Host    string `json:"host_name"`
+	Service string `json:"service_name"`
+	Author  string `json:"author"`
+	Comment string `json:"comment"`
+}
+
+// HostServiceRuntimeAttributes are common attributes of both Host and Service objects.
+//
+// When catching up potentially missed changes, the following fields seems to be the relevant ones which, fortunately,
+// are identical for Icinga 2 Host and Service objects.
+//
+// NOTE:
+//   - Name is either the Host or the Service name.
+//   - Host is empty for Host objects; Host contains the Service's Host object name for Services.
+//   - State might be 0 = UP, 1 = DOWN for hosts and 0 = OK, 1 = WARNING, 2 = CRITICAL, 3 = UNKNOWN for services.
+//   - StateType might be 0 = SOFT, 1 = HARD.
+//   - Acknowledgement type is 0 = NONE, 1 = NORMAL, 2 = STICKY.
+//
+// https://icinga.com/docs/icinga-2/latest/doc/09-object-types/#host
+// https://icinga.com/docs/icinga-2/latest/doc/09-object-types/#service
+type HostServiceRuntimeAttributes struct {
+	Name                      string      `json:"name"`
+	Host                      string      `json:"host_name,omitempty"`
+	Groups                    []string    `json:"groups"`
+	State                     int         `json:"state"`
+	StateType                 int         `json:"state_type"`
```

Comment:
This and other structs in this file have `StateType` attributes but they seem to be used nowhere. So far, only hard states where processed but the current implementation processes any state.

Reply:
@julianbrost: I did some small refactoring to reduce the number of magic integer constant values mirrored from the Icinga 2 API and introduced checks to only process HARD state changes.